{
 "cells": [
  {
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-11-25T18:49:49.663430Z",
     "start_time": "2025-11-25T18:49:47.371387Z"
    }
   },
   "cell_type": "code",
   "outputs": [],
   "execution_count": 1,
   "source": [
    "import scanpy as sc\n",
    "from scipy import io"
   ],
   "id": "ac95468d1a57838e"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:49:54.823208Z",
     "start_time": "2025-11-25T18:49:54.695889Z"
    }
   },
   "cell_type": "code",
   "source": "!mkdir matrix_files # Make new directory/folder called \"matrix_files\"",
   "id": "8ec57aca7361db45",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:52:08.071805Z",
     "start_time": "2025-11-25T18:52:05.996507Z"
    }
   },
   "cell_type": "code",
   "source": "adata = sc.read_h5ad('human_immune_health_atlas_cd8t-gdt-mait.h5ad') # Read .h5ad file and create variable adata",
   "id": "189d828387c17148",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "# adata = adata.raw.to_adata() Run the code behind arrow if working with raw data",
   "id": "b834f9281981b0e5"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:53:29.890437Z",
     "start_time": "2025-11-25T18:53:29.839752Z"
    }
   },
   "cell_type": "code",
   "source": [
    "with open('matrix_files/barcodes.tsv', 'w') as f: # Run code to create barcodes.tsv needed for Seurat (not .gz yet)\n",
    "    for item in adata.obs_names:\n",
    "        f.write(item + '\\n')"
   ],
   "id": "a3052c9c09c3a7ea",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:53:34.126901Z",
     "start_time": "2025-11-25T18:53:34.116080Z"
    }
   },
   "cell_type": "code",
   "source": [
    "with open('matrix_files/features.tsv', 'w') as f: # Run code to create features.tsv needed for Seurat (not .gz yet)\n",
    "    for item in ['\\t'.join([x,x,'Gene Expression']) for x in adata.var_names]:\n",
    "        f.write(item + '\\n')"
   ],
   "id": "c00ec9ef1e7f320e",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:53:45.510615Z",
     "start_time": "2025-11-25T18:53:37.935646Z"
    }
   },
   "cell_type": "code",
   "source": "io.mmwrite('matrix_files/matrix', adata.X.T) # Run code to create matrix.mtx needed for Seurat (not .gz yet)",
   "id": "1437be37c936bc9a",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:53:50.183741Z",
     "start_time": "2025-11-25T18:53:50.039421Z"
    }
   },
   "cell_type": "code",
   "source": "!ls matrix_files/ # Check all 3 files are present (barcodes, features, and matrix)",
   "id": "db67f12816a05eea",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcodes.tsv features.tsv matrix.mtx\r\n"
     ]
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:58:36.556153Z",
     "start_time": "2025-11-25T18:56:15.075708Z"
    }
   },
   "cell_type": "code",
   "source": "!gzip matrix_files/* # Zip all files",
   "id": "7841a38659fa6610",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:58:43.797004Z",
     "start_time": "2025-11-25T18:58:43.670478Z"
    }
   },
   "cell_type": "code",
   "source": "!ls matrix_files/ # Check all 3 files are zipped (barcodes.gz, features.gz, and matrix.gz)",
   "id": "2d2b7cd3a7a1a509",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcodes.tsv.gz features.tsv.gz matrix.mtx.gz\r\n"
     ]
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-25T18:59:09.508782Z",
     "start_time": "2025-11-25T18:59:07.395576Z"
    }
   },
   "cell_type": "code",
   "source": "adata.obs.to_csv('metadata.csv') # Create metadata file as a .csv",
   "id": "c67234d201aff170",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "# From here switch over to R Studio IDE to transform files (barcodes.gz, features.gz, and matrix.gz) to a single .h5 file",
   "id": "b3de317015ac773"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# BiocManager::install(\"DropletUtils\") Run script only first time installing DropletUtils\n",
    "library(DropletUtils)\n",
    "library(Seurat)"
   ],
   "id": "3d2d6a86a3fd84"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "matrix_dir = '/Directory' # Define the directory containing the 3 files",
   "id": "1a1e774a99e2a050"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "counts <- Read10X(data.dir = matrix_dir) # Read the data into an expression matrix. The Read10X function automatically looks for the three standard file names within the directory",
   "id": "d8c90c9819222ca9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "save_file_path <- '/Directory' # Define the path and name for the output h5 file\n",
    "save_name <- \"filtered_feature_bc_matrix\" # The final file will be called filtered_feature_bc_matrix.h5"
   ],
   "id": "c8181bf7e71e26db"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "write10xCounts( # Use write10xCounts to create the H5 file\n",
    "  file.path(save_file_path, paste0(save_name, \".h5\")),\n",
    "  counts,\n",
    "  type = \"HDF5\",\n",
    "  genome = \"GRCh38\", # Most recent (Dec,2025) human\n",
    "  version = \"3\", # Specify genome and version (e.g., \"mm10\", \"3\") as needed for your data. Version 3 is the current standard\n",
    "  overwrite = TRUE,\n",
    "  gene.id = rownames(counts),\n",
    "  gene.symbol = rownames(counts)\n",
    ")"
   ],
   "id": "bda146be3f12716d"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
